File: examples/ffi/node/discover.js
import path from "node:path";
import { fileURLToPath } from "node:url";
import koffi from "koffi";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const root = path.resolve(__dirname, "..", "..", "..");
const releaseDir = path.join(root, "target", "release");

function libraryName() {
  switch (process.platform) {
    case "darwin":
      return "liblnd.dylib";
    case "win32":
      return "lnd.dll";
    default:
      return "liblnd.so";
  }
}

const lib = koffi.load(path.join(releaseDir, libraryName()));

const lnd_client_new = lib.func("lnd_client_new", "void *", ["str", "str"]);
const lnd_client_free = lib.func("lnd_client_free", "void", ["void *"]);
const lnd_discovery_filter_new = lib.func("lnd_discovery_filter_new", "void *", ["str"]);
const lnd_discovery_filter_set_service = lib.func(
  "lnd_discovery_filter_set_service",
  "bool",
  ["void *", "str"],
);
const lnd_discovery_filter_free = lib.func("lnd_discovery_filter_free", "void", ["void *"]);
const lnd_discover = lib.func("lnd_discover", "void *", ["void *", "void *"]);
const lnd_string_free = lib.func("lnd_string_free", "void", ["void *"]);
const lnd_last_error = lib.func("lnd_last_error", "str", []);

const client = lnd_client_new("http://127.0.0.1:8765", "dev-token");
if (!client) {
  throw new Error(lnd_last_error());
}

const filter = lnd_discovery_filter_new("office-a");
if (!filter) {
  lnd_client_free(client);
  throw new Error(lnd_last_error());
}

if (!lnd_discovery_filter_set_service(filter, "_demo._tcp")) {
  lnd_discovery_filter_free(filter);
  lnd_client_free(client);
  throw new Error(lnd_last_error());
}

const jsonPtr = lnd_discover(client, filter);
if (!jsonPtr) {
  lnd_discovery_filter_free(filter);
  lnd_client_free(client);
  throw new Error(lnd_last_error());
}

console.log(koffi.decode(jsonPtr, "char", -1));

lnd_string_free(jsonPtr);
lnd_discovery_filter_free(filter);
lnd_client_free(client);

File: examples/ffi/java/Discover.java
import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Pointer;

public final class Discover {
    public interface LndLibrary extends Library {
        LndLibrary INSTANCE = Native.load("lnd", LndLibrary.class);

        Pointer lnd_client_new(String serverUrl, String bearerToken);
        void lnd_client_free(Pointer handle);
        Pointer lnd_discovery_filter_new(String networkId);
        boolean lnd_discovery_filter_set_service(Pointer handle, String service);
        void lnd_discovery_filter_free(Pointer handle);
        Pointer lnd_discover(Pointer client, Pointer filter);
        void lnd_string_free(Pointer value);
        String lnd_last_error();
    }

    public static void main(String[] args) {
        LndLibrary lib = LndLibrary.INSTANCE;
        Pointer client = lib.lnd_client_new("http://127.0.0.1:8765", "dev-token");
        if (client == null) {
            throw new IllegalStateException(lib.lnd_last_error());
        }

        Pointer filter = lib.lnd_discovery_filter_new("office-a");
        if (filter == null) {
            lib.lnd_client_free(client);
            throw new IllegalStateException(lib.lnd_last_error());
        }

        if (!lib.lnd_discovery_filter_set_service(filter, "_demo._tcp")) {
            lib.lnd_discovery_filter_free(filter);
            lib.lnd_client_free(client);
            throw new IllegalStateException(lib.lnd_last_error());
        }

        Pointer json = lib.lnd_discover(client, filter);
        if (json == null) {
            lib.lnd_discovery_filter_free(filter);
            lib.lnd_client_free(client);
            throw new IllegalStateException(lib.lnd_last_error());
        }

        System.out.println(json.getString(0));
        lib.lnd_string_free(json);
        lib.lnd_discovery_filter_free(filter);
        lib.lnd_client_free(client);
    }
}
